oak/server/rust/oak_runtime/src/node/grpc/client.rs

```diff
@@ -63,67 +67,134 @@ impl GrpcClientNode {
-    ) -> Result<(), OakStatus> {
+    async fn handle_loop(&self, runtime: RuntimeProxy, handle: Handle) -> Result<(), OakStatus> {
         // Create a [`Receiver`] used for reading gRPC invocations.
         let receiver = Receiver::<Invocation>::new(handle);
         loop {
-            debug!("Waiting for gRPC invocations");
+            debug!("Waiting for gRPC invocation");
             // Read a gRPC invocation from the [`Receiver`].
             let invocation = receiver.receive(&runtime).map_err(|error| {
                 error!("Couldn't receive the invocation: {:?}", error);
                 error
             })?;
 
-            // Receive a request from the invocation channel.
-            let request = invocation.receive_request(&runtime).map_err(|error| {
-                error!(
-                    "Couldn't receive gRPC request from the invocation: {:?}",
-                    error
-                );
+            let result = self.process_invocation(&runtime, &invocation).await;
+            info!("Invocation processing finished: {:?}", result);
+            invocation.close(&runtime);
+        }
+    }
+
+    /// Process a gRPC method invocation for an external gRPC service.
+    async fn process_invocation(
+        &self,
+        runtime: &RuntimeProxy,
+        invocation: &Invocation,
+    ) -> Result<(), OakStatus> {
+        // Receive a request from the invocation channel.
+        let request = invocation.receive_request(&runtime).map_err(|error| {
+            invocation.send_error(rpc::Code::Internal, "Failed to read request", runtime);
+            error!(
+                "Couldn't read gRPC request from the invocation: {:?}",
                 error
+            );
+            error
+        })?;
+        debug!("Incoming gRPC request: {:?}", request);
+
+        // Connect to an external gRPC service.
+        let mut grpc_client = self.connect().await.map_err(|error| {
+            error!("Couldn't connect to {}: {:?}", self.uri, error);
+            invocation.send_error(rpc::Code::NotFound, "Service connection failed", runtime);
+            OakStatus::ErrInternal
+        })?;
+        grpc_client.ready().await.map_err(|error| {
+            error!("Service was not ready: {}", error);
+            invocation.send_error(rpc::Code::NotFound, "Service not ready", runtime);
+            OakStatus::ErrInternal
+        })?;
+
+        let codec = VecCodec::default();
+        let path = request.method_name.parse().map_err(|error| {
+            error!("Invalid URI {}: {}", request.method_name, error);
+            invocation.send_error(rpc::Code::InvalidArgument, "Invalid URI", runtime);
+            OakStatus::ErrInternal
+        })?;
+
+        // Forward the request to the external gRPC service and wait for the response(s).
+        let request = tonic::Request::new(request.req_msg);
+        let request_stream =
+            request.map(|m| futures_util::stream::once(futures_util::future::ready(m)));
```

Comment:
Will it send only a single request in this stream?

Reply:
Yes, this is only adding support for server-streaming methods, not client-streaming.  I've tried to structure things to make it easier to possibly add client-streaming later, but I don't yet know how or if the async stuff will work (the code fundamentally needs to be waiting on two disparate work sources at the same time, the `wait_on_channels` for Node-triggered requests and the HTTP connection for gRPC-traffic-triggered responses).